### PR TITLE
feat: support more source for json list object

### LINKED ISSUE #605: Cannot unmarshal array into Go value of type map[string]interface {} with MQTT source
- Create below stream.
```json
{
   "sql" : "create stream demo () WITH (FORMAT=\"JSON\", DATASOURCE=\"devices/+/messages\" )"
}
```
- Create below rule.
```json
{
  "id": "rule1",
  "sql": "SELECT * FROM demo",
  "actions": [
    {
      "log": {}
    },
    {
      "mqtt": {
        "server": "tcp://broker.emqx.io:1883",
        "topic": "devices/result",
        "qos": 1,
        "clientId": "demo_001"
      }
    }
  ]
}
```
- Send with below data

```json
[
  {
    "att1" : 1,
    "att2" : 2
  },
  {
    "att1" : 3,
    "att2" : 4
  }
]
```

In the log, it print below error.
```
time="2020-11-17 12:51:41" level=error msg="Invalid data format, cannot convert [\n  {\n    \"att1\" : 1,\n    \"att2\" : 2\n  },\n  {\n    \"att1\" : 3,\n    \"att2\" : 4\n  }\n] into JSON with error json: cannot unmarshal array into Go value of type map[string]interface {}" file="extensions/mqtt_source.go:165" rule=rule1
```

### PR DESCRIPTION
**Pull request by Yisaer**
Closes #605

### CODE REVIEW

File: internal/io/http/client.go [hunk 1]
```diff
@@ -350,20 +351,22 @@
...
+		payloads, err := decode(ctx, c)
 		if err != nil {
 			return nil, c, err
 		}
-		ro := &bodyResp{}
-		err = cast.MapToStruct(payload, ro)
-		if err != nil {
-			return nil, c, fmt.Errorf("invalid body response: %v", err)
-		}
 		if resp.StatusCode < 200 || resp.StatusCode > 299 {
```
Comment (ngjaying) [inline internal/io/http/client.go:358]: This was a bug, it actually wants to check ro.Code. Some http service do not use http status code, but put the code in the payload.

**Review by ngjaying — CHANGES_REQUESTED** [review]

File: internal/io/http/client.go (hunk 1 above)
Comment (Yisaer) [reply internal/io/http/client.go]: updated.

File: internal/io/http/client.go [hunk 2]
```diff
@@ -350,20 +351,22 @@ func (cc *ClientConf) parseResponse(ctx api.StreamContext, resp *http.Response,
...
-		err = cast.MapToStruct(payload, ro)
-		if err != nil {
-			return nil, c, fmt.Errorf("invalid body response: %v", err)
-		}
-		if resp.StatusCode < 200 || resp.StatusCode > 299 {
-			return nil, c, fmt.Errorf("http status code is not 200: %v", payload)
+		for _, payload := range payloads {
+			ro := &bodyResp{}
+			err = cast.MapToStruct(payload, ro)
+			if err != nil {
```
Comment (ngjaying) [inline internal/io/http/client.go:361]: Not sure if we should fail whole response for one parsing error or still return partial result.

**Review by ngjaying — APPROVED** [review]

File: internal/io/http/client.go (hunk 2 above)
Comment (Yisaer) [reply internal/io/http/client.go:361]: I think fail is better, otherwise it may look like we discard some requests.

**Pull request merged**